### PR TITLE
Separate out strptime.pyx from tslib

### PR DESCRIPTION
**Pull request by jbrockmendel**
This is the 2nd of an N part series of PRs to split `tslib` into independent modules.

At the moment there is a big chunk of code at the bottom of `tslib` that looks like it was pasted in from somewhere else.  The header for that section of the file reads `# Don't even ask`.  So I won't.

The new `tslibs.strptime` only used in one place: `array_strptime` is called in `tools.datetimes`.  Other than that, nothing needs to be exposed, and nothing else in `tslib` relies on it.

The one function from `tslib` that `strptime` _does_ need is `_check_dts_bounds`.  This (mostly) moves that up to datetime.pxd, which both `tslib` and `strptime` already import anyway.

This is _mostly_ a copy/paste of the existing functions+classes.  I cleaned up a couple of places where variables used camelCase.

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by chris-b1** [comment]
Please don't make changes and move things in the same commit - it completely hides the changes when we're reviewing - thanks!

**Comment by jreback** [comment]
what is the rationale for moving this in the first place?

**Comment by jbrockmendel** [comment]
> what is the rationale for moving this in the first place?

In order of "most specific to this PR" to "most general to this sequence of PRs":

1) It is not immediately obvious that `strptime` is independent of the rest of `tslib`.  Separating it out makes both pieces easier to maintain.  DAGs FTW.

2) `tslib` is huge and gets imported in a whole bunch of places, often times just for `NaT`.  Making that import more light-weight is a win.

3) Over at dateutil we're talking about the next big push, and I'm advocating for making things more pandas-friendly.  e.g. see `tslib.dateutil_parse` docstring `""" lifted from dateutil to get resolution"""`.  Having dateutil return this resolution information would make this workaround unnecessary, improving performance and code complexity.
To that end, I'm trying to centralize the date parsing code that happens in a handful of places around pandas (`core.tools.datetimes`, `_libs.tslib`, `_libs/src/inference.pyx`) and separate it out into module(s) with minimal intra-pandas dependencies.

4) A big part of the goal for this series of PRs is to disentangle the dependencies between `tslib`, `period`, and `tseries`, as `tseries` is a big performance bottleneck (in particular `tseries.offsets.DateOffset`, see #17181 ).

5) See this comment in `_libs.__init__`: 

```
# TODO
# period is directly dependent on tslib and imports python
# modules, so exposing Period as an alias is currently not possible
```

One of the big goals of this sequence of PRs is handling this.

**Comment by jbrockmendel** [comment]
Just pushed a commit to remove cython "non-standard" cython decorators and revert a camelCase fix.  i.e. with the exception of a couple of flake8 whitespace changes, this should be a cut/paste of the existing functions.

**Comment by jreback** [comment]
this will need a rebase after #17422

File: pandas/_libs/src/datetime.pxd [hunk 1]
```diff
@@ -161,3 +162,23 @@ cdef inline int64_t _date_to_datetime64(object val,
     dts.hour = dts.min = dts.sec = dts.us = 0
     dts.ps = dts.as = 0
     return pandas_datetimestruct_to_datetime(PANDAS_FR_ns, dts)
+
+
+cdef inline check_dts_bounds(pandas_datetimestruct *dts):
+    cdef:
```
Comment (jreback) [inline pandas/_libs/src/datetime.pxd]: better to make this a bint, and just return True/False.

Then make another function which *calls* this one which actually raises the error.

File: pandas/_libs/tslib.pyx [hunk 2]
```diff
@@ -1734,17 +1736,11 @@ class OutOfBoundsDatetime(ValueError):
...
-
-    if dts.year <= 1677 and cmp_pandas_datetimestruct(dts, &_NS_MIN_DTS) == -1:
-        error = True
-    elif (
-            dts.year >= 2262 and
-            cmp_pandas_datetimestruct(dts, &_NS_MAX_DTS) == 1):
-        error = True
-
-    if error:
+    # Retaining this is a kludge because I haven't figured out how
```
Comment (jreback) [inline pandas/_libs/tslib.pyx]: see above

File: setup.py [hunk 3]
```diff
@@ -464,6 +464,8 @@ def pxd(name):
                    'pandas/_libs/src/period_helper.h',
                    'pandas/_libs/src/datetime.pxd']
 
+np_dtime_strs_srcs = ['pandas/_libs/src/datetime/np_datetime.c',
```
Comment (jreback) [inline setup.py]: these sources are also used else where, just list them for now directly (e.g. like period.pyx and others do it). This doesn't add anything

File: pandas/_libs/tslib.pyx [hunk 4]
```diff
@@ -5356,317 +5074,4 @@ def shift_months(int64_t[:] dtindex, int months, object day=None):
 
 #----------------------------------------------------------------------
```
Comment (jreback) [inline pandas/_libs/tslib.pyx]: you can take this comment out :<

File: pandas/_libs/tslib.pyx [hunk 5]
```diff
@@ -5356,317 +5074,4 @@ def shift_months(int64_t[:] dtindex, int months, object day=None):
...
-    week_0_length = (7 - first_weekday) % 7
-    if week_of_year == 0:
-        return 1 + day_of_week - first_weekday
-    else:
-        days_to_week = week_0_length + (7 * (week_of_year - 1))
-        return 1 + days_to_week + day_of_week
-
-# def _strptime_time(data_string, format="%a %b %d %H:%M:%S %Y"):
-#     return _strptime(data_string, format)[0]
+from tslibs.strptime import array_strptime
```
Comment (jreback) [inline pandas/_libs/tslib.pyx]: I don't think you need this at all; this is called in exactly 1 place in the code simply call it there. then strptime is decoupled.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslib.pyx (hunk 5 above)
Comment (jbrockmendel) [reply pandas/_libs/tslib.pyx]: Sounds great.  I've been importing things back into `tslib` with `#noqa` to keep the namespace unchanged.  Is that unnecessary more generally?

File: setup.py (hunk 3 above)
Comment (jbrockmendel) [reply setup.py]: OK.  This will lead to flake8 warnings that I'm happy to ignore.

File: pandas/_libs/src/datetime.pxd (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/src/datetime.pxd]: The longer-term solution that I've implemented but not PRed is to move a bunch of datetime.pxd into a pyx file where OutOfBoundsDatetime can be defined.  This also ends up clearing out _a lot_ of the setup.py dependencies (orthogonal to this discussion, but still).

Is returning a bint actually any simpler than re-raising?  It isn't obvious whether True means "error state is True" or "everything is OK is True", whereas raising is unambiguous.

That said, I'll do it your way.

File: pandas/_libs/tslib.pyx (hunk 5 above)
Comment (jreback) [reply pandas/_libs/tslib.pyx]: we might not be fully flaking this type of thing (or maybe we are and have to use noqa). I *would* like to remove non-essential imports though. These are ll private namespaces.

File: pandas/_libs/src/datetime.pxd (hunk 1 above)
Comment (jreback) [reply pandas/_libs/src/datetime.pxd]: yeah its ok, just don't like to muddle even more.

File: pandas/_libs/tslibs/strptime.pyx [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+"""Strptime-related classes and functions.
+
+CLASSES:
+    LocaleTime -- Discovers and stores locale-specific time information
+    TimeRE -- Creates regexes for pattern matching a string of text containing
+                time information
+
+FUNCTIONS:
+    _getlang -- Figure out what language is being used for the locale
+    strptime -- Calculates the time struct represented by the passed-in string
```
Comment (jorisvandenbossche) [inline pandas/_libs/tslibs/strptime.pyx]: I know you copy pasted this, but this does not really suite for the module docstring, as this is only for part of the file. 

So I would either move this inline to where those functions/classes are defined (as it was in tslib), or update it to reflect the actual file contents

If I understand correctly `array_strptime` is the only 'public' (for pandas, as used outside this file) function? If so, I would focus on that in the module docstring

**Comment by jorisvandenbossche** [comment]
There is some other code related to parsing strings in tslib as well. Does it make sense to move those as well?

**Comment by jorisvandenbossche** [comment]
> There is some other code related to parsing strings in tslib as well. Does it make sense to move those as well?

Ah, I see there is already https://github.com/pandas-dev/pandas/pull/17363 :-)

File: pandas/_libs/tslibs/strptime.pyx (hunk 6 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/strptime.pyx]: There are a handful of style issues I'd like to circle back to (e.g. mixed_camel_Case), will add this to the list.  For the time being I'm _really_ eager to wrapping this up.

File: pandas/_libs/tslibs/strptime.pyx [hunk 7]
```diff
@@ -0,0 +1,631 @@
...
+    pandas_datetimestruct_to_datetime)
+
+cimport util
+from util cimport is_string_object, get_nat
+
+cdef int64_t NPY_NAT = util.get_nat()
+
+cdef set _nat_strings = set(['NaT', 'nat', 'NAT', 'nan', 'NaN', 'NAN'])
+
+
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: can we move both of these to util.pxd? to avoid repeating code

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/strptime.pyx (hunk 7 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/strptime.pyx]: `_nat_strings` and what else?

File: pandas/_libs/tslibs/strptime.pyx (hunk 6 above)
Comment (jorisvandenbossche) [reply pandas/_libs/tslibs/strptime.pyx]: This is not really a style issue, but how you copy pasted things. If you want to clean-up later, I would just put this in the middle close of the file close to the relevant code for now (like it was in tslib)

Comment (jbrockmendel) [reply pandas/_libs/tslibs/strptime.pyx]: Not sure I understand.  Are you OK with this being addressed in a follow-up PR?

**Comment by jreback** [comment]
needs a rebase

File: pandas/_libs/tslibs/strptime.pyx (hunk 6 above)
Comment (jreback) [reply pandas/_libs/tslibs/strptime.pyx]: I would rather you fix the docs now. just move these docs down to where they are used, and put array_strptime up front

File: pandas/_libs/tslibs/strptime.pyx (hunk 7 above)
Comment (jreback) [reply pandas/_libs/tslibs/strptime.pyx]: checknull_with_nat, this seems like its repeating in lots of places.

**Review by jreback — CHANGES_REQUESTED** [review]
generally ok with this. a couple of doc comments. if you can move checknull_with_nat (to avoid repeating code all over the place) to a .pxd great. if not, can do later. but pls make an issue in that case.

**Comment by jbrockmendel** [comment]
> checknull_with_nat, this seems like its repeating in lots of places

Ultimately both `checknull_with_nat` and `_nat_strings` belong in the not-yet-introduced `tslibs.nattype`.  First choice would be to live with this bit of redundancy until that is ready.  But putting them in util.pxd (or datetime.pxd) wouldn't be the end of the world.

File: pandas/_libs/tslibs/strptime.pyx (hunk 6 above)
Comment (jorisvandenbossche) [reply pandas/_libs/tslibs/strptime.pyx]: @jbrockmendel I see you integrated them into the class/functions docstrings, which is good, but I would still include a comment where LocaleTime starts (the place where previously the comment was) to make it clear that this part is some kind of vendored code (it just comes from the standard library: https://github.com/python/cpython/blob/master/Lib/_strptime.py)

File: pandas/_libs/tslibs/strptime.pyx [hunk 8]
```diff
@@ -0,0 +1,627 @@
...
+                                                       dts.min, dts.sec)
+                raise OutOfBoundsDatetime(
+                    'Out of bounds nanosecond timestamp: %s' % fmt)
+
+    return result
+
+
+cdef inline bint _checknull_with_nat(object val):
+    """ utility to check if a value is a nat or not """
+    return (val is None or
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: move this to the top and add a TODO to consolidate in future

File: pandas/_libs/tslibs/strptime.pyx [hunk 9]
```diff
@@ -0,0 +1,627 @@
...
+    # the same as that specified by %U or %W).
+    week_0_length = (7 - first_weekday) % 7
+    if week_of_year == 0:
+        return 1 + day_of_week - first_weekday
+    else:
+        days_to_week = week_0_length + (7 * (week_of_year - 1))
+        return 1 + days_to_week + day_of_week
+
+
+# def _strptime_time(data_string, format="%a %b %d %H:%M:%S %Y"):
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: remove commented

File: pandas/_libs/tslibs/strptime.pyx [hunk 10]
```diff
@@ -0,0 +1,627 @@
...
+        dts.sec = second
+        dts.us = us
+        dts.ps = ns * 1000
+
+        iresult[i] = pandas_datetimestruct_to_datetime(PANDAS_FR_ns, &dts)
+        if check_dts_bounds(&dts):
+            if is_coerce:
+                iresult[i] = NPY_NAT
+                continue
+            else:
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: maybe worth it having a tiny extension util.pyx that houses things  like this (and checknull for example)
something that we can import anywhere that isn't a dependency itself

Comment (jreback) [reply pandas/_libs/tslibs/strptime.pyx]: actually i bet inference can be this extension

Comment (jbrockmendel) [reply pandas/_libs/tslibs/strptime.pyx]: > inference.pyx can be this extension

? inference.pyx depends on a bunch of stuff, including `tslib`.

The solution I've used locally is to put this into a `tslibs.npy_dtime` which has no dependencies.

Comment (jreback) [reply pandas/_libs/tslibs/strptime.pyx]: ok, the name needs work though :>

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

File: pandas/_libs/tslibs/strptime.pyx (hunk 10 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/strptime.pyx]: > ok, the name needs work though :>

OK, as long as it isn't another thing named `datetime`.

**Pull request merged**